runtime/compiler/optimizer/HCRGuardAnalysis.cpp

```diff
@@ -88,80 +89,148 @@ bool TR_HCRGuardAnalysis::shouldSkipBlock(TR::Block *block)
 
 void TR_HCRGuardAnalysis::initializeGenAndKillSetInfo()
    {
+   int32_t numBits = getNumberOfBits();
    for (int32_t i = 0; i < comp()->getFlowGraph()->getNextNodeNumber(); ++i)
       {
-      _regularGenSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(getNumberOfBits(),trMemory(), stackAlloc);
-      _exceptionGenSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(getNumberOfBits(),trMemory(), stackAlloc);
-      _regularKillSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(getNumberOfBits(),trMemory(), stackAlloc);
-      _exceptionKillSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(getNumberOfBits(),trMemory(), stackAlloc);
+      _regularGenSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(numBits,trMemory(), stackAlloc);
+      _exceptionGenSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(numBits,trMemory(), stackAlloc);
+      _regularKillSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(numBits,trMemory(), stackAlloc);
+      _exceptionKillSetInfo[i] = new (trStackMemory()) TR_SingleBitContainer(numBits,trMemory(), stackAlloc);
       }
 
    TR::Block *currentBlock = NULL;
-   bool exceptingTTSeen = false;
+   int32_t blockId = -1;
    TR::NodeChecklist checklist(comp());
-   TR::Node *osrNode;
+   TR::NodeChecklist exceptionNodelist(comp());
+   bool isGen, isKill, hasExceptionGen, hasExceptionKill, isYieldPoint;
 
    TR_ByteCodeInfo nodeBCI;
    nodeBCI.setCallerIndex(-1);
    nodeBCI.setByteCodeIndex(0);
    nodeBCI.setDoNotProfile(false);
    currentBlock = comp()->getStartTree()->getEnclosingBlock();
    if (!comp()->getMethodSymbol()->supportsInduceOSR(nodeBCI, currentBlock, comp()))
-      _regularGenSetInfo[currentBlock->getNumber()]->setAll(getNumberOfBits());
+      _regularGenSetInfo[currentBlock->getNumber()]->setAll(numBits);
 
    for (TR::TreeTop *treeTop = comp()->getStartTree(); treeTop; treeTop = treeTop->getNextTreeTop())
       {
-      if (treeTop->getNode()->getOpCodeValue() == TR::BBStart)
+      TR::Node* ttNode = treeTop->getNode();
+      TR::Node* osrNode = NULL;
+      isYieldPoint = false;
+      if (ttNode->getOpCodeValue() == TR::BBStart)
          {
-         exceptingTTSeen = false;
          currentBlock = treeTop->getEnclosingBlock();
+         blockId = currentBlock->getNumber();
+         isGen = isKill = hasExceptionGen = hasExceptionKill = false;
```

Comment:
I think a comment above this about resetting the walk state would be helpful. I'm not a fan of the chained equals because it makes it much harder to read what this is doing IMO